File: Leetcode/3583.py
from collections import Counter
from typing import List


class Solution:
    def specialTriplets(self, nums: List[int]) -> int:
        mod = 10**9 + 7
        arr = nums
        n = len(arr)
        prefix = Counter()
        suffix = Counter()
        for i in range(n - 1, 0, -1):
            suffix[arr[i]] += 1
        prefix[arr[0]] += 1
        ans = 0
        for j in range(1, n - 1):
            suffix[arr[j]] -= 1
            ans = ans + prefix[arr[j] << 1] * suffix[arr[j] << 1]
            if ans >= mod:
                ans %= mod
            prefix[arr[j]] += 1
        return ans

File: Leetcode/3583.go
package Leetcode

const mod = int(1e9 + 7)

func specialTriplets(nums []int) int {
	ans := 0
	n := len(nums)
	prefixCounter := make(map[int]int)
	suffixCounter := make(map[int]int)

	for i := n - 1; i > 0; i-- {
		suffixCounter[nums[i]]++
	}

	prefixCounter[nums[0]]++

	for i := 1; i < n-1; i++ {
		suffixCounter[nums[i]]--
		ans = (ans + prefixCounter[nums[i]<<1]*suffixCounter[nums[i]<<1]) % mod
		prefixCounter[nums[i]]++
	}

	return ans

}
